core/src/main/scala/cats/syntax/foldable.scala

```diff
@@ -158,4 +158,37 @@ final class FoldableOps[F[_], A](val fa: F[A]) extends AnyVal {
       case Some((a, src)) => G.map(p(a))(if (_) Right(Some(a)) else Left(src.value))
       case None => G.pure(Right(None))
     })
+
+  /**
+    * Tear down a subset of this structure using a `PartialFunction`.
+    *{{{
+    * scala> import cats.implicits._
+    * scala> val xs = List(1, 2, 3, 4)
+    * scala> xs.collectFold { case n if n % 2 == 0 => n }
+    * res0: Int = 6
+    *}}}
+    */
+  def collectFold[M: Monoid](f: PartialFunction[A, M])(implicit F: Foldable[F]): M = {
+    val m = Monoid[M]
+    F.foldLeft(fa, m.empty)((acc, a) ⇒ m.combine(acc, f.applyOrElse(a, (_: A) ⇒ m.empty)))
+  }
+
+  /**
+    * Tear down a subset of this structure using a `A => Option[M]`.
+    *{{{
+    * scala> import cats.implicits._
+    * scala> val xs = List(1, 2, 3, 4)
+    * scala> def f(n: Int): Option[Int] = if (n % 2 == 0) Some(n) else None
+    * scala> xs.filterFold(f)
+    * res0: Int = 6
+    *}}}
+    */
+  def filterFold[M: Monoid](f: A ⇒ Option[M])(implicit F: Foldable[F]): M = {
```

Comment:
We could indeed define these in terms of each other. I'm not sure if they'd be more performant than just using `foldLeft`, but if we do this, then it should be `filterFold` in terms of `collectFold`, since `A => Option[M]` has to allocate an extra `Option` :) 

Up to you, @satansk :)